# Difference method + joint learning

In this model, we will use what we called the Difference method. We first calculate the difference between the experimental band gap and the PBE band gap, and we then evaluate the error on this difference. We have three targets in the dataframe: the experimental and PBE band gaps, and the difference between them. We will apply joint learning on these columns.

In [1]:
def setup_threading():
    import os
    os.environ['OPENBLAS_NUM_THREADS'] = '1'
    os.environ['MKL_NUM_THREADS'] = '1'
    os.environ["OMP_NUM_THREADS"] = "1"
    os.environ["TF_NUM_INTRAOP_THREADS"] = "1"
    os.environ["TF_NUM_INTEROP_THREADS"] = "1"
    os.environ["CUDA_VISIBLE_DEVICES"] = "-1"

In [2]:
setup_threading()

In [3]:
from modnet.preprocessing import MODData
from modnet.models.vanilla import MODNetModel
from modnet.hyper_opt.fit_genetic import FitGenetic
import numpy as np
import os
import copy
from tensorflow.keras.callbacks import EarlyStopping, ReduceLROnPlateau

In [4]:
from sklearn.model_selection import KFold
from modnet.preprocessing import MODData

def shuffle_MD(data,random_state=10):
    data = copy.deepcopy(data)
    ids = data.df_targets.sample(frac=1,random_state=random_state).index
    data.df_featurized = data.df_featurized.reindex([ids])
    data.df_targets = data.df_targets.reindex([ids])
    data.df_structure = data.df_structure.reindex([ids])
    
    return data

def MDKsplit(data,n_splits=5,random_state=10):
    #data = shuffle_MD(data,random_state=random_state)
    ids = np.array(data.structure_ids)
    kf = KFold(n_splits=n_splits,shuffle=True,random_state=random_state)
    folds = []
    for train_idx, val_idx in kf.split(ids):
        data_train = MODData(data.df_structure.iloc[train_idx]['structure'].values,data.df_targets.iloc[train_idx].values,target_names=data.df_targets.columns,structure_ids=ids[train_idx])
        data_train.df_featurized = data.df_featurized.iloc[train_idx]
        #data_train.optimal_features = data.optimal_features
        
        data_val = MODData(data.df_structure.iloc[val_idx]['structure'].values,data.df_targets.iloc[val_idx].values,target_names=data.df_targets.columns,structure_ids=ids[val_idx])
        data_val.df_featurized = data.df_featurized.iloc[val_idx]
        #data_val.optimal_features = data.optimal_features

        folds.append((data_train,data_val))
        
    return folds

def MD_append(md,lmd):
    md = copy.deepcopy(md)
    for m in lmd:
        md.df_structure.append(m.df_structure)
        md.df_targets.append(m.df_targets)
        md.df_featurized.append(m.df_featurized)
    return md

In [3]:
#from random import randint
#def train_val_split(data):

#    """Splits arrays or matrices into random train and validation subsets.
#
#   Parameter:
#      data: 'MODData' data which need to be splitted.
#    """
    #i = randint(0,9)
    #md_train, md_val = MDKsplit(data, n_splits=5, random_state=1)[0]
    #y_train = md_train.df_targets
    #y_val = md_val.df_targets

    #return md_train, md_val, y_train, y_val

In [4]:
#train_val_split(md_joint)

In [5]:
md_exp = MODData.load('exp_gap_all_mpid')
md_exp.df_targets.columns = ['exp_gap','mp_id']
md_pbe = MODData.load('pbe_gap.zip')
md_pbe.df_targets.columns = ['gap']
md_joint = MODData.load('exp_pbe_joint')


If you use the ChemEnv tool for your research, please consider citing the following reference(s) :
David Waroquiers, Xavier Gonze, Gian-Marco Rignanese, Cathrin Welker-Nieuwoudt, Frank Rosowski,
Michael Goebel, Stephan Schenk, Peter Degelmann, Rute Andre, Robert Glaum, and Geoffroy Hautier,
"Statistical analysis of coordination environments in oxides",
Chem. Mater., 2017, 29 (19), pp 8346-8360,
DOI: 10.1021/acs.chemmater.7b02766

2021-05-05 11:51:24,381 - modnet - INFO - Loaded <modnet.preprocessing.MODData object at 0x7f54f42e8700> object, created with modnet version <=0.1.7
2021-05-05 11:51:31,026 - modnet - INFO - Loaded <modnet.preprocessing.MODData object at 0x7f54f42eeac0> object, created with modnet version <=0.1.7
2021-05-05 11:51:31,050 - modnet - INFO - Loaded <modnet.preprocessing.MODData object at 0x7f54f42da760> object, created with modnet version <=0.1.7


In [6]:
md_joint.df_targets

exp_gap  pbe_gap
mp-12699      0.00   0.0000
mp-559459     3.40   2.1704
mp-21162      0.00   0.0000
mp-1306       0.00   0.0000
mp-15252      0.53   0.1049
...            ...      ...
mp-556541     1.80   1.5135
mp-2602       1.04   0.0000
mp-30366      0.00   0.0000
mp-3718       1.45   0.2418
mp-3709       0.00   0.0000

[1703 rows x 2 columns]

In [7]:
md_joint.df_targets['difference'] = md_joint.df_targets['exp_gap'] - md_joint.df_targets['pbe_gap']

In [8]:
md_joint.df_targets

exp_gap  pbe_gap  difference
mp-12699      0.00   0.0000      0.0000
mp-559459     3.40   2.1704      1.2296
mp-21162      0.00   0.0000      0.0000
mp-1306       0.00   0.0000      0.0000
mp-15252      0.53   0.1049      0.4251
...            ...      ...         ...
mp-556541     1.80   1.5135      0.2865
mp-2602       1.04   0.0000      1.0400
mp-30366      0.00   0.0000      0.0000
mp-3718       1.45   0.2418      1.2082
mp-3709       0.00   0.0000      0.0000

[1703 rows x 3 columns]

In [9]:
import pymatgen.ext.matproj

In [10]:
from pymatgen.ext.matproj import MPRester
import pandas as pd

with MPRester("v454AlrxL5Zg0cLAXW0") as mpr:
    index = md_joint.df_structure.index
    array = np.empty(shape=md_joint.df_structure.shape, dtype=pymatgen.core.structure.Structure)
    array = np.ndarray.flatten(array)
    for i in range(len(md_joint.df_structure)):
        array[i] = mpr.get_structure_by_material_id(index[i])
    md_joint.df_structure = pd.DataFrame(data=array, columns=["structure"])
    md_joint.df_structure = md_joint.df_structure.set_index(index)

In [11]:
#md_joint.save('exp_pbe_joint')

2021-05-05 11:58:44,955 - modnet - INFO - Data successfully saved as exp_pbe_joint!


In [5]:
md_joint = MODData.load('exp_pbe_joint')


If you use the ChemEnv tool for your research, please consider citing the following reference(s) :
David Waroquiers, Xavier Gonze, Gian-Marco Rignanese, Cathrin Welker-Nieuwoudt, Frank Rosowski,
Michael Goebel, Stephan Schenk, Peter Degelmann, Rute Andre, Robert Glaum, and Geoffroy Hautier,
"Statistical analysis of coordination environments in oxides",
Chem. Mater., 2017, 29 (19), pp 8346-8360,
DOI: 10.1021/acs.chemmater.7b02766

2021-05-20 15:06:30,044 - modnet - INFO - Loaded <modnet.preprocessing.MODData object at 0x7f6a8c4d4d00> object, created with modnet version <=0.1.7


In [6]:
md_joint.df_structure

structure
id                                                          
mp-12699   [[ 0.13616363 -2.60631371  1.50340165] Ca, [-1...
mp-559459  [[2.14049282 6.43464803 1.76199846] Cs, [-0.70...
mp-21162   [[0. 0. 0.] Sr, [0.        2.5259875 2.5259875...
mp-1306    [[3.259129   1.88166243 1.15971025] Th, [ 3.25...
mp-15252   [[0.95531657 1.042239   0.86229681] Cu, [2.803...
...                                                      ...
mp-556541  [[0.         6.49363948 5.82657217] In, [ 0.  ...
mp-2602     [[3.023318 3.023318 3.023318] Nd, [0. 0. 0.] As]
mp-30366   [[4.51774558 1.37529211 4.32659015] Ca, [1.109...
mp-3718    [[2.0697935  0.         1.28932533] Th, [0.   ...
mp-3709    [[0. 0. 0.] Tb, [0.       1.983905 2.537711] C...

[1703 rows x 1 columns]

In [7]:
from sklearn.metrics import mean_absolute_error
mean_absolute_error(md_joint.df_targets['exp_gap'], md_joint.df_targets['pbe_gap'])

0.429783975697596

In [8]:
md_joint.df_targets['difference'].mad()

0.5494667780829975

In [9]:
import uncertainty_toolbox as uct

k = 5
random_state = 202010
folds = MDKsplit(md_joint,n_splits=k,random_state=random_state)
errors = np.ones((int(len(md_joint.df_targets)/k),5))
maes = np.ones(5)
uncertainties = np.ones(5)
metricss = []
for i,f in enumerate(folds):
    train = f[0]
    test = f[1]
    #train.feature_selection(n=-1, use_precomputed_cross_nmi=True)
    fpath = 'train_joint_exp-pbe_{}_{}'.format(random_state,i+1)
    if os.path.exists(fpath):
        train = MODData.load(fpath)
    else:
        train.feature_selection(n=-1, use_precomputed_cross_nmi=True)
        train.save(fpath)
       
    # assure no overlap
    assert len(set(train.df_targets.index).intersection(set(test.df_targets.index))) == 0
    
    # difference prediction
    ga = FitGenetic(train)
    model = ga.run(size_pop=20, num_generations=10, n_jobs=20)
    
    pred, std = model.predict(test, return_unc=True)
    true = test.df_targets
    true = true.drop(columns=['exp_gap','pbe_gap'])
    metrics = uct.metrics.get_all_metrics(pred['difference'].values, std['difference'].values, true['difference'].values)
    error = pred-true
    error = error['difference'].drop(pred.index[((pred['difference']).abs()>20)]) # drop unrealistic values: happens extremely rarely
    std = std['difference'].drop(std.index[((std['difference']).abs()>20)]) # drop unrealistic values: happens extremely rarely
    mae = np.abs(error.values).mean()
    uncertainty = np.abs(std).mean()
    print('MAE =')
    print(mae)
    print('uncertainty =')
    print(uncertainty)
    maes[i] = mae
    uncertainties[i] = uncertainty
    metricss.append(metrics)

2021-05-20 15:06:34,511 - modnet - INFO - Loaded DeBreuck2020Featurizer featurizer.
2021-05-20 15:06:34,556 - modnet - INFO - Loaded DeBreuck2020Featurizer featurizer.
2021-05-20 15:06:34,593 - modnet - INFO - Loaded DeBreuck2020Featurizer featurizer.
2021-05-20 15:06:34,644 - modnet - INFO - Loaded DeBreuck2020Featurizer featurizer.
2021-05-20 15:06:34,681 - modnet - INFO - Loaded DeBreuck2020Featurizer featurizer.
2021-05-20 15:06:34,734 - modnet - INFO - Loaded DeBreuck2020Featurizer featurizer.
2021-05-20 15:06:34,769 - modnet - INFO - Loaded DeBreuck2020Featurizer featurizer.
2021-05-20 15:06:34,817 - modnet - INFO - Loaded DeBreuck2020Featurizer featurizer.
2021-05-20 15:06:34,853 - modnet - INFO - Loaded DeBreuck2020Featurizer featurizer.
2021-05-20 15:06:34,904 - modnet - INFO - Loaded DeBreuck2020Featurizer featurizer.
2021-05-20 15:06:35,365 - modnet - INFO - Loaded <modnet.preprocessing.MODData object at 0x7f6a19a8a430> object, created with modnet version 0.1.9
2021-05-20 15

100%|██████████| 100/100 [13:37<00:00,  8.18s/it]

2021-05-20 15:20:16,412 - modnet - INFO - Loss per individual: ind 0: 0.867 	ind 1: 0.973 	ind 2: 0.921 	ind 3: 1.068 	ind 4: 0.879 	ind 5: 0.938 	ind 6: 0.903 	ind 7: 0.850 	ind 8: 0.951 	ind 9: 0.894 	ind 10: 0.826 	ind 11: 0.863 	ind 12: 0.964 	ind 13: 0.881 	ind 14: 0.864 	ind 15: 0.901 	ind 16: 0.885 	ind 17: 0.981 	ind 18: 0.859 	ind 19: 0.858 	


2021-05-20 15:20:19,512 - modnet - INFO - Generation number 1
2021-05-20 15:20:23,076 - modnet - INFO - Multiprocessing on 20 cores. Total of 32 cores available.


100%|██████████| 100/100 [11:02<00:00,  6.63s/it] 

2021-05-20 15:31:25,828 - modnet - INFO - Loss per individual: ind 0: 0.893 	ind 1: 0.999 	ind 2: 0.876 	ind 3: 1.159 	ind 4: 0.941 	ind 5: 0.857 	ind 6: 0.874 	ind 7: 0.964 	ind 8: 0.845 	ind 9: 0.934 	ind 10: 0.903 	ind 11: 0.915 	ind 12: 0.900 	ind 13: 0.823 	ind 14: 0.851 	ind 15: 0.886 	ind 16: 0.844 	ind 17: 0.974 	ind 18: 0.856 	ind 19: 1.099 	


2021-05-20 15:31:28,480 - modnet - INFO - Generation number 2
2021-05-20 15:31:32,433 - modnet - INFO - Multiprocessing on 20 cores. Total of 32 cores available.


100%|██████████| 100/100 [12:35<00:00,  7.56s/it]

2021-05-20 15:44:08,377 - modnet - INFO - Loss per individual: ind 0: 0.861 	ind 1: 0.860 	ind 2: 0.856 	ind 3: 0.828 	ind 4: 0.940 	ind 5: 0.888 	ind 6: 0.811 	ind 7: 0.912 	ind 8: 0.871 	ind 9: 0.914 	ind 10: 0.856 	ind 11: 0.877 	ind 12: 0.887 	ind 13: 0.880 	ind 14: 0.918 	ind 15: 0.978 	ind 16: 0.857 	ind 17: 0.895 	ind 18: 0.966 	ind 19: 0.900 	


2021-05-20 15:44:11,116 - modnet - INFO - Generation number 3
2021-05-20 15:44:15,413 - modnet - INFO - Multiprocessing on 20 cores. Total of 32 cores available.


100%|██████████| 100/100 [09:43<00:00,  5.84s/it]

2021-05-20 15:53:59,169 - modnet - INFO - Loss per individual: ind 0: 0.849 	ind 1: 0.924 	ind 2: 0.876 	ind 3: 0.850 	ind 4: 0.847 	ind 5: 0.998 	ind 6: 0.843 	ind 7: 0.869 	ind 8: 0.958 	ind 9: 0.838 	ind 10: 0.857 	ind 11: 0.957 	ind 12: 0.840 	ind 13: 0.840 	ind 14: 0.972 	ind 15: 0.837 	ind 16: 0.905 	ind 17: 0.837 	ind 18: 0.833 	ind 19: 0.877 	


2021-05-20 15:54:01,724 - modnet - INFO - Generation number 4
2021-05-20 15:54:06,409 - modnet - INFO - Multiprocessing on 20 cores. Total of 32 cores available.


100%|██████████| 100/100 [08:35<00:00,  5.16s/it]

2021-05-20 16:02:42,528 - modnet - INFO - Loss per individual: ind 0: 0.843 	ind 1: 0.854 	ind 2: 0.852 	ind 3: 0.956 	ind 4: 0.906 	ind 5: 1.009 	ind 6: 0.887 	ind 7: 0.834 	ind 8: 0.865 	ind 9: 0.866 	ind 10: 0.840 	ind 11: 0.968 	ind 12: 0.898 	ind 13: 0.826 	ind 14: 0.853 	ind 15: 0.897 	ind 16: 1.005 	ind 17: 0.889 	ind 18: 0.877 	ind 19: 0.862 	


2021-05-20 16:02:45,076 - modnet - INFO - Generation number 5
2021-05-20 16:02:50,844 - modnet - INFO - Multiprocessing on 20 cores. Total of 32 cores available.


100%|██████████| 100/100 [12:47<00:00,  7.67s/it]

2021-05-20 16:15:38,130 - modnet - INFO - Loss per individual: ind 0: 0.874 	ind 1: 0.864 	ind 2: 0.918 	ind 3: 0.824 	ind 4: 0.842 	ind 5: 0.903 	ind 6: 0.853 	ind 7: 0.848 	ind 8: 0.931 	ind 9: 0.844 	ind 10: 0.869 	ind 11: 0.831 	ind 12: 0.906 	ind 13: 0.850 	ind 14: 0.924 	ind 15: 0.876 	ind 16: 0.870 	ind 17: 0.918 	ind 18: 0.901 	ind 19: 0.823 	


2021-05-20 16:15:41,018 - modnet - INFO - Early stopping: same best model for 4 consecutive generations
 (1/n) Calculating accuracy metrics
 (2/n) Calculating average calibration metrics


  0%|          | 0/10 [00:00<?, ?it/s]

 (3/n) Calculating adversarial group calibration metrics
  [1/2] for mean absolute calibration error
Measuring adversarial group calibration by spanning group size between 0.0 and 1.0, in 10 intervals


  0%|          | 0/10 [00:00<?, ?it/s]

  [2/2] for root mean squared calibration error
Measuring adversarial group calibration by spanning group size between 0.0 and 1.0, in 10 intervals


100%|██████████| 10/10 [00:02<00:00,  4.22it/s]


 (4/n) Calculating sharpness metrics
 (n/n) Calculating proper scoring rule metrics
**Finished Calculating All Metrics**


===================== Accuracy Metrics =====================
  MAE           0.332
  RMSE          0.678
  MDAE          0.066
  MARPD         146.903
  R2            0.399
  Correlation   0.647
=============== Average Calibration Metrics ================
  Root-mean-squared Calibration Error   0.112
  Mean-absolute Calibration Error       0.088
  Miscalibration Area                   0.089
========== Adversarial Group Calibration Metrics ===========
  Mean-absolute Adversarial Group Calibration Error
     Group Size: 0.11 -- Calibration Error: 0.168
     Group Size: 0.56 -- Calibration Error: 0.114
     Group Size: 1.00 -- Calibration Error: 0.088
  Root-mean-squared Adversarial Group Calibration Error
     Group Size: 0.11 -- Calibration Error: 0.195
     Group Size: 0.56 -- Calibration Error: 0.135
     Group Size: 1.00 -- Calibration Error: 0.112
==============

100%|██████████| 100/100 [11:10<00:00,  6.70s/it]

2021-05-20 16:27:13,984 - modnet - INFO - Loss per individual: ind 0: 8.358 	ind 1: 90.506 	ind 2: 0.964 	ind 3: 0.955 	ind 4: 18.738 	ind 5: 31.249 	ind 6: 0.947 	ind 7: 1.404 	ind 8: 29.454 	ind 9: 11.318 	ind 10: 4.528 	ind 11: 7.596 	ind 12: 0.985 	ind 13: 58.669 	ind 14: 3.540 	ind 15: 25.821 	ind 16: 2.446 	ind 17: 36.225 	ind 18: 7.595 	ind 19: 66.070 	


2021-05-20 16:27:16,613 - modnet - INFO - Generation number 1
2021-05-20 16:27:21,690 - modnet - INFO - Multiprocessing on 20 cores. Total of 32 cores available.


100%|██████████| 100/100 [07:16<00:00,  4.36s/it]

2021-05-20 16:34:37,967 - modnet - INFO - Loss per individual: ind 0: 18.620 	ind 1: 0.958 	ind 2: 0.966 	ind 3: 18.338 	ind 4: 0.945 	ind 5: 7.221 	ind 6: 20.713 	ind 7: 0.953 	ind 8: 0.932 	ind 9: 0.971 	ind 10: 2.005 	ind 11: 2.389 	ind 12: 0.998 	ind 13: 0.977 	ind 14: 5.164 	ind 15: 0.967 	ind 16: 2.612 	ind 17: 131.715 	ind 18: 0.944 	ind 19: 0.978 	


2021-05-20 16:34:40,965 - modnet - INFO - Generation number 2
2021-05-20 16:34:46,242 - modnet - INFO - Multiprocessing on 20 cores. Total of 32 cores available.


100%|██████████| 100/100 [07:06<00:00,  4.27s/it]

2021-05-20 16:41:53,107 - modnet - INFO - Loss per individual: ind 0: 0.923 	ind 1: 0.984 	ind 2: 14.635 	ind 3: 1.007 	ind 4: 0.959 	ind 5: 3.355 	ind 6: 30.416 	ind 7: 1.002 	ind 8: 0.948 	ind 9: 26.463 	ind 10: 0.947 	ind 11: 105.312 	ind 12: 0.946 	ind 13: 0.958 	ind 14: 0.946 	ind 15: 1.440 	ind 16: 3.076 	ind 17: 0.933 	ind 18: 0.917 	ind 19: 5.790 	


2021-05-20 16:41:55,674 - modnet - INFO - Generation number 3
2021-05-20 16:42:01,649 - modnet - INFO - Multiprocessing on 20 cores. Total of 32 cores available.


100%|██████████| 100/100 [08:26<00:00,  5.07s/it]

2021-05-20 16:50:28,418 - modnet - INFO - Loss per individual: ind 0: 0.945 	ind 1: 0.931 	ind 2: 0.945 	ind 3: 0.978 	ind 4: 0.979 	ind 5: 1.008 	ind 6: 1.019 	ind 7: 0.934 	ind 8: 0.995 	ind 9: 5.088 	ind 10: 1.853 	ind 11: 0.953 	ind 12: 0.925 	ind 13: 1.437 	ind 14: 0.917 	ind 15: 0.913 	ind 16: 0.914 	ind 17: 0.935 	ind 18: 7.878 	ind 19: 1.323 	


2021-05-20 16:50:31,258 - modnet - INFO - Generation number 4
2021-05-20 16:50:36,570 - modnet - INFO - Multiprocessing on 20 cores. Total of 32 cores available.


100%|██████████| 100/100 [09:14<00:00,  5.55s/it]

2021-05-20 16:59:51,420 - modnet - INFO - Loss per individual: ind 0: 11.750 	ind 1: 0.948 	ind 2: 0.930 	ind 3: 0.927 	ind 4: 0.963 	ind 5: 2.188 	ind 6: 0.930 	ind 7: 0.921 	ind 8: 0.919 	ind 9: 0.931 	ind 10: 12.096 	ind 11: 0.929 	ind 12: 0.964 	ind 13: 0.946 	ind 14: 0.917 	ind 15: 0.935 	ind 16: 0.948 	ind 17: 0.936 	ind 18: 0.968 	ind 19: 1.139 	


2021-05-20 16:59:53,696 - modnet - INFO - Generation number 5
2021-05-20 17:00:00,015 - modnet - INFO - Multiprocessing on 20 cores. Total of 32 cores available.


100%|██████████| 100/100 [06:56<00:00,  4.16s/it]

2021-05-20 17:06:56,310 - modnet - INFO - Loss per individual: ind 0: 0.945 	ind 1: 0.956 	ind 2: 0.935 	ind 3: 0.936 	ind 4: 0.951 	ind 5: 0.986 	ind 6: 1.006 	ind 7: 0.968 	ind 8: 0.960 	ind 9: 0.939 	ind 10: 1.150 	ind 11: 0.917 	ind 12: 5.984 	ind 13: 0.939 	ind 14: 0.912 	ind 15: 0.946 	ind 16: 0.919 	ind 17: 1.108 	ind 18: 1.005 	ind 19: 0.914 	


2021-05-20 17:06:59,350 - modnet - INFO - Generation number 6
2021-05-20 17:07:05,307 - modnet - INFO - Multiprocessing on 20 cores. Total of 32 cores available.


100%|██████████| 100/100 [08:21<00:00,  5.01s/it]

2021-05-20 17:15:26,848 - modnet - INFO - Loss per individual: ind 0: 0.935 	ind 1: 0.966 	ind 2: 1.171 	ind 3: 0.952 	ind 4: 0.955 	ind 5: 0.936 	ind 6: 0.940 	ind 7: 0.940 	ind 8: 0.973 	ind 9: 0.933 	ind 10: 0.956 	ind 11: 0.983 	ind 12: 0.909 	ind 13: 0.971 	ind 14: 0.918 	ind 15: 0.976 	ind 16: 0.996 	ind 17: 0.969 	ind 18: 0.956 	ind 19: 0.945 	


2021-05-20 17:15:29,781 - modnet - INFO - Generation number 7
2021-05-20 17:15:36,873 - modnet - INFO - Multiprocessing on 20 cores. Total of 32 cores available.


100%|██████████| 100/100 [07:23<00:00,  4.43s/it]

2021-05-20 17:23:00,248 - modnet - INFO - Loss per individual: ind 0: 0.914 	ind 1: 2.279 	ind 2: 0.948 	ind 3: 0.976 	ind 4: 1.021 	ind 5: 0.954 	ind 6: 0.958 	ind 7: 0.946 	ind 8: 0.916 	ind 9: 0.928 	ind 10: 0.942 	ind 11: 1.077 	ind 12: 0.914 	ind 13: 1.999 	ind 14: 0.927 	ind 15: 0.956 	ind 16: 0.940 	ind 17: 0.956 	ind 18: 0.954 	ind 19: 0.926 	


2021-05-20 17:23:02,803 - modnet - INFO - Generation number 8
2021-05-20 17:23:09,689 - modnet - INFO - Multiprocessing on 20 cores. Total of 32 cores available.


100%|██████████| 100/100 [07:48<00:00,  4.68s/it]

2021-05-20 17:30:58,101 - modnet - INFO - Loss per individual: ind 0: 0.997 	ind 1: 0.922 	ind 2: 0.950 	ind 3: 0.931 	ind 4: 0.932 	ind 5: 1.777 	ind 6: 0.904 	ind 7: 0.946 	ind 8: 0.945 	ind 9: 1.825 	ind 10: 0.987 	ind 11: 2.532 	ind 12: 0.942 	ind 13: 0.977 	ind 14: 0.936 	ind 15: 0.987 	ind 16: 1.100 	ind 17: 4.027 	ind 18: 0.926 	ind 19: 0.931 	


2021-05-20 17:31:00,710 - modnet - INFO - Generation number 9
2021-05-20 17:31:08,151 - modnet - INFO - Multiprocessing on 20 cores. Total of 32 cores available.


100%|██████████| 100/100 [07:56<00:00,  4.77s/it]

2021-05-20 17:39:04,867 - modnet - INFO - Loss per individual: ind 0: 0.955 	ind 1: 0.945 	ind 2: 0.910 	ind 3: 0.969 	ind 4: 0.966 	ind 5: 0.970 	ind 6: 0.965 	ind 7: 0.903 	ind 8: 0.928 	ind 9: 2.428 	ind 10: 0.944 	ind 11: 0.921 	ind 12: 1.006 	ind 13: 0.933 	ind 14: 0.925 	ind 15: 0.941 	ind 16: 41.620 	ind 17: 0.928 	ind 18: 0.970 	ind 19: 0.941 	


 (1/n) Calculating accuracy metrics
 (2/n) Calculating average calibration metrics


  0%|          | 0/10 [00:00<?, ?it/s]

 (3/n) Calculating adversarial group calibration metrics
  [1/2] for mean absolute calibration error
Measuring adversarial group calibration by spanning group size between 0.0 and 1.0, in 10 intervals


  0%|          | 0/10 [00:00<?, ?it/s]

  [2/2] for root mean squared calibration error
Measuring adversarial group calibration by spanning group size between 0.0 and 1.0, in 10 intervals


100%|██████████| 10/10 [00:02<00:00,  4.62it/s]


 (4/n) Calculating sharpness metrics
 (n/n) Calculating proper scoring rule metrics
**Finished Calculating All Metrics**


===================== Accuracy Metrics =====================
  MAE           0.285
  RMSE          0.632
  MDAE          0.029
  MARPD         148.770
  R2            0.390
  Correlation   0.626
=============== Average Calibration Metrics ================
  Root-mean-squared Calibration Error   0.151
  Mean-absolute Calibration Error       0.139
  Miscalibration Area                   0.141
========== Adversarial Group Calibration Metrics ===========
  Mean-absolute Adversarial Group Calibration Error
     Group Size: 0.11 -- Calibration Error: 0.197
     Group Size: 0.56 -- Calibration Error: 0.157
     Group Size: 1.00 -- Calibration Error: 0.139
  Root-mean-squared Adversarial Group Calibration Error
     Group Size: 0.11 -- Calibration Error: 0.219
     Group Size: 0.56 -- Calibration Error: 0.168
     Group Size: 1.00 -- Calibration Error: 0.151
==============

100%|██████████| 100/100 [13:55<00:00,  8.36s/it]

2021-05-20 17:53:33,435 - modnet - INFO - Loss per individual: ind 0: 0.890 	ind 1: 3.017 	ind 2: 0.855 	ind 3: 1.279 	ind 4: 1.290 	ind 5: 0.896 	ind 6: 1.025 	ind 7: 0.960 	ind 8: 1.701 	ind 9: 0.919 	ind 10: 0.916 	ind 11: 0.942 	ind 12: 3.094 	ind 13: 1.455 	ind 14: 0.975 	ind 15: 0.896 	ind 16: 0.972 	ind 17: 0.949 	ind 18: 0.878 	ind 19: 0.919 	


2021-05-20 17:53:35,767 - modnet - INFO - Generation number 1
2021-05-20 17:53:43,602 - modnet - INFO - Multiprocessing on 20 cores. Total of 32 cores available.


100%|██████████| 100/100 [11:37<00:00,  6.97s/it]

2021-05-20 18:05:20,833 - modnet - INFO - Loss per individual: ind 0: 0.943 	ind 1: 0.863 	ind 2: 0.894 	ind 3: 0.902 	ind 4: 0.959 	ind 5: 0.936 	ind 6: 0.933 	ind 7: 0.914 	ind 8: 0.915 	ind 9: 0.942 	ind 10: 0.959 	ind 11: 1.225 	ind 12: 0.877 	ind 13: 0.963 	ind 14: 0.852 	ind 15: 0.936 	ind 16: 0.869 	ind 17: 0.917 	ind 18: 0.895 	ind 19: 0.955 	


2021-05-20 18:05:23,533 - modnet - INFO - Generation number 2
2021-05-20 18:05:30,004 - modnet - INFO - Multiprocessing on 20 cores. Total of 32 cores available.


100%|██████████| 100/100 [14:05<00:00,  8.45s/it]

2021-05-20 18:19:35,623 - modnet - INFO - Loss per individual: ind 0: 0.898 	ind 1: 1.001 	ind 2: 0.853 	ind 3: 0.923 	ind 4: 0.934 	ind 5: 0.874 	ind 6: 1.016 	ind 7: 0.940 	ind 8: 0.961 	ind 9: 0.917 	ind 10: 0.932 	ind 11: 0.947 	ind 12: 0.962 	ind 13: 0.914 	ind 14: 1.162 	ind 15: 0.882 	ind 16: 0.909 	ind 17: 0.912 	ind 18: 0.960 	ind 19: 0.922 	


2021-05-20 18:19:38,628 - modnet - INFO - Generation number 3
2021-05-20 18:19:45,332 - modnet - INFO - Multiprocessing on 20 cores. Total of 32 cores available.


100%|██████████| 100/100 [11:36<00:00,  6.96s/it]

2021-05-20 18:31:21,816 - modnet - INFO - Loss per individual: ind 0: 0.966 	ind 1: 0.891 	ind 2: 0.964 	ind 3: 0.943 	ind 4: 0.885 	ind 5: 0.952 	ind 6: 8.672 	ind 7: 0.897 	ind 8: 0.917 	ind 9: 0.958 	ind 10: 0.932 	ind 11: 0.911 	ind 12: 0.938 	ind 13: 0.910 	ind 14: 0.927 	ind 15: 0.859 	ind 16: 0.924 	ind 17: 0.960 	ind 18: 0.885 	ind 19: 0.918 	


2021-05-20 18:31:24,054 - modnet - INFO - Generation number 4
2021-05-20 18:31:31,765 - modnet - INFO - Multiprocessing on 20 cores. Total of 32 cores available.


100%|██████████| 100/100 [11:52<00:00,  7.12s/it]

2021-05-20 18:43:24,030 - modnet - INFO - Loss per individual: ind 0: 0.909 	ind 1: 0.949 	ind 2: 0.901 	ind 3: 0.987 	ind 4: 0.927 	ind 5: 0.947 	ind 6: 0.863 	ind 7: 0.927 	ind 8: 0.869 	ind 9: 0.914 	ind 10: 0.954 	ind 11: 0.996 	ind 12: 0.957 	ind 13: 1.117 	ind 14: 0.878 	ind 15: 0.929 	ind 16: 0.938 	ind 17: 0.874 	ind 18: 0.897 	ind 19: 0.912 	


2021-05-20 18:43:26,641 - modnet - INFO - Early stopping: same best model for 4 consecutive generations
 (1/n) Calculating accuracy metrics
 (2/n) Calculating average calibration metrics


  0%|          | 0/10 [00:00<?, ?it/s]

 (3/n) Calculating adversarial group calibration metrics
  [1/2] for mean absolute calibration error
Measuring adversarial group calibration by spanning group size between 0.0 and 1.0, in 10 intervals


  0%|          | 0/10 [00:00<?, ?it/s]

  [2/2] for root mean squared calibration error
Measuring adversarial group calibration by spanning group size between 0.0 and 1.0, in 10 intervals


100%|██████████| 10/10 [00:02<00:00,  4.37it/s]


 (4/n) Calculating sharpness metrics
 (n/n) Calculating proper scoring rule metrics
**Finished Calculating All Metrics**


===================== Accuracy Metrics =====================
  MAE           0.319
  RMSE          0.697
  MDAE          0.018
  MARPD         150.064
  R2            0.451
  Correlation   0.675
=============== Average Calibration Metrics ================
  Root-mean-squared Calibration Error   0.110
  Mean-absolute Calibration Error       0.094
  Miscalibration Area                   0.095
========== Adversarial Group Calibration Metrics ===========
  Mean-absolute Adversarial Group Calibration Error
     Group Size: 0.11 -- Calibration Error: 0.161
     Group Size: 0.56 -- Calibration Error: 0.116
     Group Size: 1.00 -- Calibration Error: 0.094
  Root-mean-squared Adversarial Group Calibration Error
     Group Size: 0.11 -- Calibration Error: 0.178
     Group Size: 0.56 -- Calibration Error: 0.127
     Group Size: 1.00 -- Calibration Error: 0.110
==============

100%|██████████| 100/100 [11:35<00:00,  6.96s/it]

2021-05-20 18:55:22,647 - modnet - INFO - Loss per individual: ind 0: 0.922 	ind 1: 1.024 	ind 2: 1.045 	ind 3: 1.081 	ind 4: 1.107 	ind 5: 0.926 	ind 6: 0.976 	ind 7: 1.136 	ind 8: 0.915 	ind 9: 0.968 	ind 10: 0.996 	ind 11: 0.991 	ind 12: 1.067 	ind 13: 0.886 	ind 14: 1.007 	ind 15: 1.154 	ind 16: 0.974 	ind 17: 0.952 	ind 18: 0.997 	ind 19: 0.932 	


2021-05-20 18:55:25,473 - modnet - INFO - Generation number 1
2021-05-20 18:55:33,303 - modnet - INFO - Multiprocessing on 20 cores. Total of 32 cores available.


100%|██████████| 100/100 [10:13<00:00,  6.13s/it]

2021-05-20 19:05:46,625 - modnet - INFO - Loss per individual: ind 0: 0.989 	ind 1: 0.971 	ind 2: 0.941 	ind 3: 1.023 	ind 4: 1.041 	ind 5: 1.155 	ind 6: 0.912 	ind 7: 0.907 	ind 8: 0.965 	ind 9: 0.905 	ind 10: 0.944 	ind 11: 1.131 	ind 12: 1.048 	ind 13: 1.001 	ind 14: 1.158 	ind 15: 0.919 	ind 16: 1.110 	ind 17: 0.934 	ind 18: 0.992 	ind 19: 1.019 	


2021-05-20 19:05:48,973 - modnet - INFO - Generation number 2
2021-05-20 19:05:56,593 - modnet - INFO - Multiprocessing on 20 cores. Total of 32 cores available.


100%|██████████| 100/100 [10:28<00:00,  6.28s/it]

2021-05-20 19:16:25,135 - modnet - INFO - Loss per individual: ind 0: 0.940 	ind 1: 0.902 	ind 2: 0.968 	ind 3: 0.952 	ind 4: 1.165 	ind 5: 1.014 	ind 6: 0.960 	ind 7: 0.922 	ind 8: 1.001 	ind 9: 0.906 	ind 10: 1.027 	ind 11: 0.929 	ind 12: 0.965 	ind 13: 0.952 	ind 14: 0.999 	ind 15: 0.963 	ind 16: 0.945 	ind 17: 0.913 	ind 18: 1.011 	ind 19: 0.902 	


2021-05-20 19:16:27,888 - modnet - INFO - Generation number 3
2021-05-20 19:16:34,655 - modnet - INFO - Multiprocessing on 20 cores. Total of 32 cores available.


100%|██████████| 100/100 [10:12<00:00,  6.12s/it]

2021-05-20 19:26:47,236 - modnet - INFO - Loss per individual: ind 0: 0.988 	ind 1: 0.941 	ind 2: 1.042 	ind 3: 0.981 	ind 4: 1.038 	ind 5: 0.966 	ind 6: 0.913 	ind 7: 0.974 	ind 8: 0.993 	ind 9: 0.923 	ind 10: 1.067 	ind 11: 1.022 	ind 12: 0.934 	ind 13: 0.907 	ind 14: 0.919 	ind 15: 0.907 	ind 16: 0.910 	ind 17: 0.962 	ind 18: 0.966 	ind 19: 0.986 	


2021-05-20 19:26:50,014 - modnet - INFO - Early stopping: same best model for 4 consecutive generations
 (1/n) Calculating accuracy metrics
 (2/n) Calculating average calibration metrics


  0%|          | 0/10 [00:00<?, ?it/s]

 (3/n) Calculating adversarial group calibration metrics
  [1/2] for mean absolute calibration error
Measuring adversarial group calibration by spanning group size between 0.0 and 1.0, in 10 intervals


  0%|          | 0/10 [00:00<?, ?it/s]

  [2/2] for root mean squared calibration error
Measuring adversarial group calibration by spanning group size between 0.0 and 1.0, in 10 intervals


100%|██████████| 10/10 [00:02<00:00,  4.49it/s]


 (4/n) Calculating sharpness metrics
 (n/n) Calculating proper scoring rule metrics
**Finished Calculating All Metrics**


===================== Accuracy Metrics =====================
  MAE           0.283
  RMSE          0.690
  MDAE          0.018
  MARPD         152.243
  R2            0.271
  Correlation   0.559
=============== Average Calibration Metrics ================
  Root-mean-squared Calibration Error   0.079
  Mean-absolute Calibration Error       0.064
  Miscalibration Area                   0.064
========== Adversarial Group Calibration Metrics ===========
  Mean-absolute Adversarial Group Calibration Error
     Group Size: 0.11 -- Calibration Error: 0.123
     Group Size: 0.56 -- Calibration Error: 0.081
     Group Size: 1.00 -- Calibration Error: 0.064
  Root-mean-squared Adversarial Group Calibration Error
     Group Size: 0.11 -- Calibration Error: 0.142
     Group Size: 0.56 -- Calibration Error: 0.099
     Group Size: 1.00 -- Calibration Error: 0.079
==============

100%|██████████| 100/100 [10:10<00:00,  6.10s/it]

2021-05-20 19:37:19,154 - modnet - INFO - Loss per individual: ind 0: 1.024 	ind 1: 1.173 	ind 2: 1.143 	ind 3: 1.027 	ind 4: 1.055 	ind 5: 1.062 	ind 6: 1.039 	ind 7: 1.120 	ind 8: 1.534 	ind 9: 1.142 	ind 10: 1.028 	ind 11: 0.973 	ind 12: 0.975 	ind 13: 1.086 	ind 14: 0.966 	ind 15: 1.084 	ind 16: 1.012 	ind 17: 1.193 	ind 18: 1.042 	ind 19: 1.030 	


2021-05-20 19:37:21,777 - modnet - INFO - Generation number 1
2021-05-20 19:37:28,017 - modnet - INFO - Multiprocessing on 20 cores. Total of 32 cores available.


100%|██████████| 100/100 [11:59<00:00,  7.19s/it]

2021-05-20 19:49:27,318 - modnet - INFO - Loss per individual: ind 0: 1.066 	ind 1: 1.104 	ind 2: 1.074 	ind 3: 0.993 	ind 4: 1.018 	ind 5: 1.101 	ind 6: 1.058 	ind 7: 1.039 	ind 8: 1.101 	ind 9: 1.055 	ind 10: 1.110 	ind 11: 0.978 	ind 12: 1.043 	ind 13: 0.976 	ind 14: 1.040 	ind 15: 1.081 	ind 16: 1.029 	ind 17: 0.975 	ind 18: 1.082 	ind 19: 1.003 	


2021-05-20 19:49:29,956 - modnet - INFO - Generation number 2
2021-05-20 19:49:36,478 - modnet - INFO - Multiprocessing on 20 cores. Total of 32 cores available.


100%|██████████| 100/100 [11:38<00:00,  6.98s/it]

2021-05-20 20:01:15,007 - modnet - INFO - Loss per individual: ind 0: 1.023 	ind 1: 1.094 	ind 2: 1.026 	ind 3: 1.037 	ind 4: 0.990 	ind 5: 1.062 	ind 6: 0.948 	ind 7: 1.024 	ind 8: 1.078 	ind 9: 1.072 	ind 10: 0.995 	ind 11: 1.117 	ind 12: 1.009 	ind 13: 1.050 	ind 14: 0.995 	ind 15: 1.098 	ind 16: 1.149 	ind 17: 1.076 	ind 18: 1.013 	ind 19: 1.237 	


2021-05-20 20:01:17,774 - modnet - INFO - Generation number 3
2021-05-20 20:01:24,975 - modnet - INFO - Multiprocessing on 20 cores. Total of 32 cores available.


100%|██████████| 100/100 [08:32<00:00,  5.12s/it]

2021-05-20 20:09:57,464 - modnet - INFO - Loss per individual: ind 0: 0.958 	ind 1: 1.142 	ind 2: 1.038 	ind 3: 1.011 	ind 4: 1.205 	ind 5: 1.020 	ind 6: 0.988 	ind 7: 0.957 	ind 8: 1.006 	ind 9: 1.020 	ind 10: 1.047 	ind 11: 0.978 	ind 12: 1.006 	ind 13: 0.968 	ind 14: 1.371 	ind 15: 0.988 	ind 16: 1.144 	ind 17: 1.261 	ind 18: 0.970 	ind 19: 0.968 	


2021-05-20 20:10:00,193 - modnet - INFO - Generation number 4
2021-05-20 20:10:08,252 - modnet - INFO - Multiprocessing on 20 cores. Total of 32 cores available.


100%|██████████| 100/100 [11:35<00:00,  6.95s/it]

2021-05-20 20:21:43,817 - modnet - INFO - Loss per individual: ind 0: 0.944 	ind 1: 1.000 	ind 2: 0.957 	ind 3: 0.988 	ind 4: 0.996 	ind 5: 0.946 	ind 6: 1.132 	ind 7: 0.985 	ind 8: 1.042 	ind 9: 1.004 	ind 10: 1.075 	ind 11: 1.025 	ind 12: 0.984 	ind 13: 1.005 	ind 14: 1.189 	ind 15: 1.119 	ind 16: 0.976 	ind 17: 0.968 	ind 18: 1.093 	ind 19: 1.380 	


2021-05-20 20:21:46,798 - modnet - INFO - Generation number 5
2021-05-20 20:21:53,918 - modnet - INFO - Multiprocessing on 20 cores. Total of 32 cores available.


100%|██████████| 100/100 [07:44<00:00,  4.64s/it]

2021-05-20 20:29:38,346 - modnet - INFO - Loss per individual: ind 0: 1.013 	ind 1: 1.178 	ind 2: 0.989 	ind 3: 1.029 	ind 4: 1.021 	ind 5: 0.958 	ind 6: 1.020 	ind 7: 1.040 	ind 8: 0.972 	ind 9: 1.049 	ind 10: 0.991 	ind 11: 1.073 	ind 12: 1.056 	ind 13: 0.923 	ind 14: 1.154 	ind 15: 1.021 	ind 16: 1.095 	ind 17: 1.439 	ind 18: 1.022 	ind 19: 1.166 	


2021-05-20 20:29:40,926 - modnet - INFO - Generation number 6
2021-05-20 20:29:49,011 - modnet - INFO - Multiprocessing on 20 cores. Total of 32 cores available.


100%|██████████| 100/100 [09:12<00:00,  5.53s/it]

2021-05-20 20:39:01,673 - modnet - INFO - Loss per individual: ind 0: 0.952 	ind 1: 1.008 	ind 2: 1.009 	ind 3: 1.126 	ind 4: 1.007 	ind 5: 1.027 	ind 6: 1.228 	ind 7: 1.057 	ind 8: 1.106 	ind 9: 0.998 	ind 10: 0.957 	ind 11: 1.033 	ind 12: 0.981 	ind 13: 1.023 	ind 14: 0.961 	ind 15: 1.023 	ind 16: 1.032 	ind 17: 0.947 	ind 18: 0.974 	ind 19: 1.038 	


2021-05-20 20:39:03,762 - modnet - INFO - Generation number 7
2021-05-20 20:39:12,478 - modnet - INFO - Multiprocessing on 20 cores. Total of 32 cores available.


100%|██████████| 100/100 [08:14<00:00,  4.95s/it]

2021-05-20 20:47:27,455 - modnet - INFO - Loss per individual: ind 0: 1.035 	ind 1: 1.117 	ind 2: 1.011 	ind 3: 0.939 	ind 4: 1.033 	ind 5: 1.149 	ind 6: 0.998 	ind 7: 0.972 	ind 8: 1.020 	ind 9: 1.099 	ind 10: 1.058 	ind 11: 1.085 	ind 12: 1.052 	ind 13: 1.032 	ind 14: 0.991 	ind 15: 1.014 	ind 16: 0.994 	ind 17: 1.077 	ind 18: 1.040 	ind 19: 1.014 	


2021-05-20 20:47:29,792 - modnet - INFO - Generation number 8
2021-05-20 20:47:38,471 - modnet - INFO - Multiprocessing on 20 cores. Total of 32 cores available.


100%|██████████| 100/100 [09:15<00:00,  5.56s/it]

2021-05-20 20:56:54,312 - modnet - INFO - Loss per individual: ind 0: 0.948 	ind 1: 0.947 	ind 2: 1.097 	ind 3: 0.956 	ind 4: 1.038 	ind 5: 1.282 	ind 6: 1.004 	ind 7: 1.001 	ind 8: 1.057 	ind 9: 1.010 	ind 10: 0.960 	ind 11: 1.007 	ind 12: 1.022 	ind 13: 1.146 	ind 14: 1.033 	ind 15: 0.940 	ind 16: 1.008 	ind 17: 1.003 	ind 18: 1.124 	ind 19: 0.936 	


2021-05-20 20:56:56,665 - modnet - INFO - Early stopping: same best model for 4 consecutive generations
 (1/n) Calculating accuracy metrics
 (2/n) Calculating average calibration metrics


  0%|          | 0/10 [00:00<?, ?it/s]

 (3/n) Calculating adversarial group calibration metrics
  [1/2] for mean absolute calibration error
Measuring adversarial group calibration by spanning group size between 0.0 and 1.0, in 10 intervals


  0%|          | 0/10 [00:00<?, ?it/s]

  [2/2] for root mean squared calibration error
Measuring adversarial group calibration by spanning group size between 0.0 and 1.0, in 10 intervals


100%|██████████| 10/10 [00:02<00:00,  4.66it/s]

 (4/n) Calculating sharpness metrics
 (n/n) Calculating proper scoring rule metrics
**Finished Calculating All Metrics**


===================== Accuracy Metrics =====================
  MAE           0.186
  RMSE          0.410
  MDAE          0.010
  MARPD         155.946
  R2            0.496
  Correlation   0.719
=============== Average Calibration Metrics ================
  Root-mean-squared Calibration Error   0.084
  Mean-absolute Calibration Error       0.070
  Miscalibration Area                   0.070
========== Adversarial Group Calibration Metrics ===========
  Mean-absolute Adversarial Group Calibration Error
     Group Size: 0.11 -- Calibration Error: 0.112
     Group Size: 0.56 -- Calibration Error: 0.082
     Group Size: 1.00 -- Calibration Error: 0.070
  Root-mean-squared Adversarial Group Calibration Error
     Group Size: 0.11 -- Calibration Error: 0.139
     Group Size: 0.56 -- Calibration Error: 0.100
     Group Size: 1.00 -- Calibration Error: 0.084
==============

In [10]:
maes

array([0.33222279, 0.28476676, 0.31888674, 0.28262384, 0.18559184])

In [11]:
maes.mean()

0.2808183948155837

In [12]:
uncertainties

array([0.12359382, 0.14253537, 0.15704803, 0.14324321, 0.11988797])

In [13]:
uncertainties.mean()

0.13726168125867844

In [14]:
metricss

[{'accuracy': {'mae': 0.33222279051059017,
   'rmse': 0.6776748256732111,
   'mdae': 0.0662896976470948,
   'marpd': 146.9026570534816,
   'r2': 0.39913545425355534,
   'corr': 0.6472028851931801},
  'avg_calibration': {'rms_cal': 0.11160391760585173,
   'ma_cal': 0.08823721081785602,
   'miscal_area': 0.08911120047164932},
  'adv_group_calibration': {'ma_adv_group_cal': {'group_sizes': array([0.        , 0.11111111, 0.22222222, 0.33333333, 0.44444444,
           0.55555556, 0.66666667, 0.77777778, 0.88888889, 1.        ]),
    'adv_group_cali_mean': array([0.43464646, 0.16839208, 0.13844949, 0.12402578, 0.11737287,
           0.11417893, 0.10807097, 0.10096897, 0.09919762, 0.08823721]),
    'adv_group_cali_stderr': array([5.94006196e-02, 2.81425271e-02, 1.60758724e-02, 1.66081478e-02,
           8.17050025e-03, 1.13574325e-02, 6.94458359e-03, 5.16110525e-03,
           4.17516084e-03, 1.46284728e-17])},
   'rms_adv_group_cal': {'group_sizes': array([0.        , 0.11111111, 0.22222222,

#### Conclusion

Relative percentage change of the error compared to MODNet_exp_GA reference: 19%